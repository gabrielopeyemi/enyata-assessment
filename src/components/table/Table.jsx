import React from 'react'
import { useNavigate } from 'react-router-dom';
import Moment from 'react-moment'

export default function Table({title, tableHeader, films, headers}) {
    const navigate = useNavigate();

    const handleNavigateToSingle = (id, url) =>{
        if(title === 'Films') navigate(`/overview/${id}`)
        if(title === 'People') navigate(`/people/${id}`)
        if(title === 'Starships') navigate(`/starships/${url.split("/")[5]}`)
        if(title === 'Species') navigate(`/species/${id}`)
    } 
    return (
        <React.Fragment>
            <h2 className=' text-coolgray text-md mb-[31px] '>{title}</h2>
            <div className="relative overflow-x-auto shadow-md sm:rounded-lg border border-antiFlashWhite">
                <table className="w-full text-sm text-left text-gray-500 bg-white">
                    <thead className="text-xs text-gray-700 uppercase bg-gray-50">
                        <tr>
                            {tableHeader.map((table, index) => (
                                <React.Fragment key={index}>
                                    {index === 0 ? (
                                        <th scope="col" className="p-4">
                                            <div className="flex items-center">
                                                <input id="checkbox-all-search" type="checkbox" className="w-4 h-4 text-blue-600 bg-gray-100 border-gray-300 rounded focus:ring-blue-500 focus:ring-2" />
                                                <label className="sr-only">{table}</label>
                                            </div>
                                        </th>
                                    ):(
                                        <th scope="col" className="px-2 capitalize text-coolgray text-md font-normal mx-2 whitespace-nowrap">
                                            {table}
                                        </th>
                                    )}

                                </React.Fragment>
                            ))}
                        </tr>
                    </thead>
                    <tbody> 
                        {films.map((film, index) => (
                            <tr onClick={() => handleNavigateToSingle(index + 1, film.url)} className="bg-white hover:bg-antiFlashWhite border-b cursor-pointer border-platinum dark:bg-gray-800 dark:border-gray-700 hover:bg-gray-50 dark:hover:bg-gray-600" key={index}>
                                <td className="w-4 p-4">
                                    <div className="flex items-center">
                                        <input id="checkbox-table-search-1" type="checkbox" className="w-4 h-4 text-blue-600 bg-gray-100 border-gray-300 rounded focus:ring-blue-500 focus:ring-2 " />
                                        <label className="sr-only">checkbox</label>
                                    </div>
                                </td>
                                {headers.map((header) => (
                                    <td className="px-2 py-4 whitespace-nowrap" key={header}>
                                        {
                                            header === 'created' ? (<Moment format="M/D/YY">{film[header]}</Moment>) :
                                            header === 'average_height' ? (<React.Fragment>{film[header]} CM</React.Fragment>) :
                                            (<React.Fragment>{film[header]}</React.Fragment>)
                                        }
                                    </td>
                                ))}
                            </tr>
                        ))}
                    </tbody>
                </table>
            </div>
        </React.Fragment>

    )
}