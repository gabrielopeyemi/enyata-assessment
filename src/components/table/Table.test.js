import React from 'react';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Table from './Table';

describe('Table component', () => {
    const tableHeader = ['Header 1', 'Header 2', 'Header 3'];
    const headers = ['name', 'created', 'average_height'];
    const films = [
        {
            name: 'Film 1',
            created: '2022-01-01T00:00:00.000Z',
            average_height: 180,
            url: 'https://swapi.dev/api/films/1/',
        },
        {
            name: 'Film 2',
            created: '2022-02-01T00:00:00.000Z',
            average_height: 175,
            url: 'https://swapi.dev/api/films/2/',
        },
    ];

    it('renders snapshot correctly', () => {
        const { asFragment } = render(
            <MemoryRouter>
                <Table title="Films" tableHeader={tableHeader} films={films} headers={headers} />
            </MemoryRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('displays correct number of rows', () => {
        render(
            <MemoryRouter>
                <Table title="Films" tableHeader={tableHeader} films={films} headers={headers} />
            </MemoryRouter>
        );
        const rows = screen.getAllByRole('row');
        expect(rows.length).toBe(films.length + 1);
    });
});
