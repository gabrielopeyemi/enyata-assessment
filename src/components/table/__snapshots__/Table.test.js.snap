// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component renders snapshot correctly 1`] = `
<DocumentFragment>
  <h2
    class=" text-coolgray text-md mb-[31px] "
  >
    Films
  </h2>
  <div
    class="relative overflow-x-auto shadow-md sm:rounded-lg border border-antiFlashWhite"
  >
    <table
      class="w-full text-sm text-left text-gray-500 bg-white"
    >
      <thead
        class="text-xs text-gray-700 uppercase bg-gray-50"
      >
        <tr>
          <th
            class="p-4"
            scope="col"
          >
            <div
              class="flex items-center"
            >
              <input
                class="w-4 h-4 text-blue-600 bg-gray-100 border-gray-300 rounded focus:ring-blue-500 focus:ring-2"
                id="checkbox-all-search"
                type="checkbox"
              />
              <label
                class="sr-only"
              >
                Header 1
              </label>
            </div>
          </th>
          <th
            class="px-2 capitalize text-coolgray text-md font-normal mx-2 whitespace-nowrap"
            scope="col"
          >
            Header 2
          </th>
          <th
            class="px-2 capitalize text-coolgray text-md font-normal mx-2 whitespace-nowrap"
            scope="col"
          >
            Header 3
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="bg-white hover:bg-antiFlashWhite border-b cursor-pointer border-platinum dark:bg-gray-800 dark:border-gray-700 hover:bg-gray-50 dark:hover:bg-gray-600"
        >
          <td
            class="w-4 p-4"
          >
            <div
              class="flex items-center"
            >
              <input
                class="w-4 h-4 text-blue-600 bg-gray-100 border-gray-300 rounded focus:ring-blue-500 focus:ring-2 "
                id="checkbox-table-search-1"
                type="checkbox"
              />
              <label
                class="sr-only"
              >
                checkbox
              </label>
            </div>
          </td>
          <td
            class="px-2 py-4 whitespace-nowrap"
          >
            Film 1
          </td>
          <td
            class="px-2 py-4 whitespace-nowrap"
          >
            <time
              datetime="1640995200000"
            >
              1/1/22
            </time>
          </td>
          <td
            class="px-2 py-4 whitespace-nowrap"
          >
            180 CM
          </td>
        </tr>
        <tr
          class="bg-white hover:bg-antiFlashWhite border-b cursor-pointer border-platinum dark:bg-gray-800 dark:border-gray-700 hover:bg-gray-50 dark:hover:bg-gray-600"
        >
          <td
            class="w-4 p-4"
          >
            <div
              class="flex items-center"
            >
              <input
                class="w-4 h-4 text-blue-600 bg-gray-100 border-gray-300 rounded focus:ring-blue-500 focus:ring-2 "
                id="checkbox-table-search-1"
                type="checkbox"
              />
              <label
                class="sr-only"
              >
                checkbox
              </label>
            </div>
          </td>
          <td
            class="px-2 py-4 whitespace-nowrap"
          >
            Film 2
          </td>
          <td
            class="px-2 py-4 whitespace-nowrap"
          >
            <time
              datetime="1643673600000"
            >
              2/1/22
            </time>
          </td>
          <td
            class="px-2 py-4 whitespace-nowrap"
          >
            175 CM
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
